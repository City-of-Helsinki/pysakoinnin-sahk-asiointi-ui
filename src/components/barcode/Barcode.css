.barcode-container {
  max-width: 90vw;
  margin-bottom: var(--spacing-m);
}

.barcode {
  overflow-wrap: break-word;
}
