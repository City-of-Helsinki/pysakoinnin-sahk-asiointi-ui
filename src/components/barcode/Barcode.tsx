import React, { useState } from 'react';
import { Button, IconCopy, Notification, TextInput } from 'hds-react';

import './Barcode.css';
import { useTranslation } from 'react-i18next';

type BarcodeProps = {
  barcode: string;
};

const Barcode = (props: BarcodeProps) => {
  const { barcode } = props;
  const { t } = useTranslation();
  const [copySuccess, setCopySuccess] = useState(false);

  const copyToClipboard = () => {
    navigator.clipboard.writeText(barcode);
    setCopySuccess(true);
  };

  return (
    <>
      <div className="barcode-container">
        <TextInput
          id="barCode"
          label={t('common:barcode:label')}
          defaultValue={barcode}
          readOnly
        />
        <Button
          className="barcode-button"
          iconLeft={<IconCopy />}
          onClick={copyToClipboard}
          variant="secondary">
          {t('common:barcode:copy-barcode')}
        </Button>
      </div>

      {copySuccess && (
        <Notification
          label={t('common:barcode:toast-label')}
          autoClose
          autoCloseDuration={5000}
          onClose={() => setCopySuccess(false)}
          type="success"
          position="top-right"
          notificationAriaLabel={t('common:barcode:toast-body')}>
          {t('common:barcode:toast-body')}
        </Notification>
      )}
    </>
  );
};

export default Barcode;
