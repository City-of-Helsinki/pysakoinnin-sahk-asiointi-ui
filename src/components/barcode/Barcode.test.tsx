import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import Barcode from './Barcode';
import { axe } from 'vitest-axe';

describe('Component', () => {
  it('matches snapshot', () => {
    const { container } = render(<Barcode barcode="test-123" />);
    expect(container).toMatchSnapshot();
  });

  it('passes a11y validation', async () => {
    const { container } = render(<Barcode barcode="test-123" />);

    expect(await axe(container)).toHaveNoViolations();
  });

  it('copies barcode to clipboard', async () => {
    Object.assign(window.navigator, {
      clipboard: {
        writeText: vi.fn().mockImplementation(() => Promise.resolve())
      }
    });

    const testBarcode = 'test-123';
    render(<Barcode barcode={testBarcode} />);

    const button = await screen.findByTestId('copy-to-clipboard');
    fireEvent.click(button);

    expect(window.navigator.clipboard.writeText).toHaveBeenCalledWith(
      testBarcode
    );
  });
});
