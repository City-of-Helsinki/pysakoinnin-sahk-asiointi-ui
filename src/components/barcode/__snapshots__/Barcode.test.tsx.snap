// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component matches snapshot 1`] = `
<div>
  <div
    class="barcode-container"
  >
    <div
      class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="barCode"
      >
        Virtuaaliviivakoodi
      </label>
      <div
        class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
      >
        <input
          class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
          id="barCode"
          readonly=""
          type="text"
          value="test-123"
        />
      </div>
    </div>
    <button
      class="Button-module_button__1msFE button_hds-button__2A0je Button-module_secondary__1nABp button_hds-button--secondary__1NOWS barcode-button"
      data-testid="copy-to-clipboard"
      type="button"
    >
      <div
        aria-hidden="true"
        class="Button-module_icon__O-h7R button_hds-icon__17j8Z"
      >
        <svg
          class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <rect
              height="24"
              width="24"
            />
            <path
              d="M6,10 L6,12 L5,12 L5,18 L12,18 L12,17 L14,17 L14,19 C14,19.5522847 13.5522847,20 13,20 L4,20 C3.44771525,20 3,19.5522847 3,19 L3,11 C3,10.4477153 3.44771525,10 4,10 L6,10 Z M20,4 C20.5522847,4 21,4.44771525 21,5 L21,15 C21,15.5522847 20.5522847,16 20,16 L8,16 C7.44771525,16 7,15.5522847 7,15 L7,5 C7,4.44771525 7.44771525,4 8,4 L20,4 Z M19,6 L9,6 L9,14 L19,14 L19,6 Z"
              fill="currentColor"
            />
          </g>
        </svg>
      </div>
      <span
        class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
      >
        Kopioi virtuaaliviivakoodi
      </span>
    </button>
  </div>
</div>
`;
